Fix WTF::HashSet<Class> and WTF::HashSet<RetainPtr<>> to work with ARC and upstream clang

ddkilzer opened the pull request:
#### f61eefa7dc8c4fa3576a0d10d06f523b7e907e9b
<pre>
Fix WTF::HashSet&lt;Class&gt; and WTF::HashSet&lt;RetainPtr&lt;&gt;&gt; to work with ARC and upstream clang
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=278624">https://bugs.webkit.org/show_bug.cgi?id=278624</a>&gt;
&lt;<a href="https://rdar.apple.com/134591290">rdar://134591290</a>&gt;

Reviewed by Darin Adler.

Upstream clang makes Class objects participate in ARC, so
WTF::HashSet&lt;Class&gt; must be changed to WTF::HashSet&lt;RetainPtr&lt;Class&gt;&gt;.

However, WTF::HashSet&lt;RetainPtr&lt;&gt;&gt; doesn&apos;t currently work with ARC, so
that had to be fixed with casting changes to WTF::RetainPtr&lt;&gt;.

Also introduce IPC::AllowedClassHashSet to use a consistent type
everywhere (except WebCore::SerializedPlatformDataCueMac).

* Source/WTF/wtf/RetainPtr.h:
(WTF::RetainPtr::RetainPtr):
(WTF::RetainPtr::isHashTableDeletedValue const):
- Use toStorageType() to fix compilation with ARC.
(WTF::RetainPtr::hashTableDeletedValue):
- Use fromStorageType() with reinterpret_cast&lt;CFTypeRef&gt;() to create the
  deleted value.
(WTF::RetainPtr::fromStorageTypeHelper):
(WTF::RetainPtr::fromStorageType):
(WTF::RetainPtr::toStorageType):
- Change to static so they can be used from hashTableDeletedValue().

* Source/WebCore/platform/mac/SerializedPlatformDataCueMac.h:
* Source/WebCore/platform/mac/SerializedPlatformDataCueMac.mm:
(WebCore::SerializedPlatformDataCueMac::allowedClassesForNativeValues):
- Switch from WTF::HashSet&lt;Class&gt; to WTF::HashSet&lt;RetainPtr&lt;Class&gt;&gt; for
  future ARC conversion and to eliminate raw pointer use.

* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::AllowedClassHashSet): Add.
- Define common type for HashSet&lt;RetainPtr&lt;Class&gt;&gt; and use in WebKit.
  This replaces HashSet&lt;Class&gt; since that no longer works with ARC.
(IPC::Decoder::decodeWithAllowedClasses):
(IPC::Decoder::allowedClasses):
- Make use of AllowedClassHashSet type.
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
(IPC::isObjectClassAllowed):
- Make use of AllowedClassHashSet type, and update for loop to work with
  the new type.
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::shouldEnableStrictMode):
(IPC::decodeObjectDirectlyRequiringAllowedClasses&lt;NSObject&lt;NSSecureCoding&gt;&gt;):
- Make use of AllowedClassHashSet type, and update for loop to work with
  the new type.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
- Add RetainPtrHashingCocoa[ARC].mm to the project.
* Tools/TestWebKitAPI/Tests/WTF/cf/RetainPtrHashing.cpp:
(TestWebKitAPI::TEST(RetainPtrHashing, HashSet)):
(TestWebKitAPI::TEST(RetainPtrHashing, HashMapKey)):
(TestWebKitAPI::TEST(RetainPtrHashing, HashMapValue)):
- Add more expectations to tests and fix header sort order.
* Tools/TestWebKitAPI/Tests/WTF/cocoa/RetainPtrHashingCocoa.mm: Copied from Tools/TestWebKitAPI/Tests/WTF/cf/RetainPtrHashing.cpp.
(TestWebKitAPI::TEST(RetainPtrHashingCocoa, HashSet)):
(TestWebKitAPI::TEST(RetainPtrHashingCocoa, HashMapKey)):
(TestWebKitAPI::TEST(RetainPtrHashingCocoa, HashMapValue)):
- Add tests for HashMap&lt;RetainPtr&lt;&gt;&gt; and HashSet&lt;RetainPtr&lt;&gt;&gt;.
* Tools/TestWebKitAPI/Tests/WTF/cocoa/RetainPtrHashingCocoaARC.mm: Add.
- Add tests for HashMap&lt;RetainPtr&lt;&gt;&gt; and HashSet&lt;RetainPtr&lt;&gt;&gt; under ARC.

Canonical link: <a href="https://commits.webkit.org/283028@main">https://commits.webkit.org/283028@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/917cef58a07025fe90ab1c0ed9abc3c6b0d18e3b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/64960 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44327 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17574 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/68984 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15566 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67078 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52110 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/15848 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52187 "Passed tests") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/10746 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68026 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/40989 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56221 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/32809 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37660 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13592 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14442 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/58073 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59571 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/13930 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70689 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/64203 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/8912 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13409 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/59516 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/8944 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56281 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/59727 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7352 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1026 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/85969 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9851 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40139 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15154 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41216 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42397 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/40960 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/mac/SerializedPlatformDataCueMac.h:
```diff
@@ -46,7 +46,7 @@ class SerializedPlatformDataCueMac final : public SerializedPlatformDataCue {
     WEBCORE_EXPORT SerializedPlatformDataCueValue encodableValue() const final;
 
-    WEBCORE_EXPORT static const HashSet<Class>& allowedClassesForNativeValues();
+    WEBCORE_EXPORT static const HashSet<RetainPtr<Class>>& allowedClassesForNativeValues();
```
Do classes need to be retained and released? If so, this is great, but if not I want to make sure we are not adding any wasteful additional work.

This seems to take this pull request from "make a future compiler happy" to something even more valuable.

ddkilzer replied:
This WebCore class change is identical to the one in WebKit, but it is forward-looking as WebCore isn't compiled with ARC yet.  Also, this removes another use of raw pointers in a `HashSet<>`.

I can remove the changes to `SerializedPlatformDataCueValue`, but it's (at minimum) inconsistent.

darinadler replied:
Seems fine to change it.

darinadler reviewed (approved):
The tests don’t test if the object lifetimes are correct and they are deallocated at the right moment and no sooner. Seems like that’s one kind of mistake we could make in the implementation.

ddkilzer commented on Source/WebKit/Platform/IPC/Decoder.h:
```diff
@@ -181,8 +181,8 @@ class Decoder {
-    HashSet<ClassStructPtr> m_allowedClasses;
+#ifdef __OBJC__
+    AllowedClassHashSet m_allowedClasses;
 #endif
```
Welp, this was the cause of all the layout test crashes.

Turns out that only defining an instance variable in Objective-C++ sources (but not C++ sources) results in a "bad time" when an object is passed between code in those two source types.

New PR coming soon.

Commit:
Fix WTF::HashSet<Class> and WTF::HashSet<RetainPtr<>> to work with ARC and upstream clang
<https://bugs.webkit.org/show_bug.cgi?id=278624>
<rdar://134591290>

Reviewed by Darin Adler.

Upstream clang makes Class objects participate in ARC, so
WTF::HashSet<Class> must be changed to WTF::HashSet<RetainPtr<Class>>.

However, WTF::HashSet<RetainPtr<>> doesn't currently work with ARC, so
that had to be fixed with casting changes to WTF::RetainPtr<>.

Also introduce IPC::AllowedClassHashSet to use a consistent type
everywhere (except WebCore::SerializedPlatformDataCueMac).

* Source/WTF/wtf/RetainPtr.h:
(WTF::RetainPtr::RetainPtr):
(WTF::RetainPtr::isHashTableDeletedValue const):
- Use toStorageType() to fix compilation with ARC.
(WTF::RetainPtr::hashTableDeletedValue):
- Use fromStorageType() with reinterpret_cast<CFTypeRef>() to create the
  deleted value.
(WTF::RetainPtr::fromStorageTypeHelper):
(WTF::RetainPtr::fromStorageType):
(WTF::RetainPtr::toStorageType):
- Change to static so they can be used from hashTableDeletedValue().

* Source/WebCore/platform/mac/SerializedPlatformDataCueMac.h:
* Source/WebCore/platform/mac/SerializedPlatformDataCueMac.mm:
(WebCore::SerializedPlatformDataCueMac::allowedClassesForNativeValues):
- Switch from WTF::HashSet<Class> to WTF::HashSet<RetainPtr<Class>> for
  future ARC conversion and to eliminate raw pointer use.

* Source/WebKit/Platform/IPC/Decoder.h:
(IPC::AllowedClassHashSet): Add.
- Define common type for HashSet<RetainPtr<Class>> and use in WebKit.
  This replaces HashSet<Class> since that no longer works with ARC.
(IPC::Decoder::decodeWithAllowedClasses):
(IPC::Decoder::allowedClasses):
- Make use of AllowedClassHashSet type.
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.h:
(IPC::isObjectClassAllowed):
- Make use of AllowedClassHashSet type, and update for loop to work with
  the new type.
* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(IPC::shouldEnableStrictMode):
(IPC::decodeObjectDirectlyRequiringAllowedClasses<NSObject<NSSecureCoding>>):
- Make use of AllowedClassHashSet type, and update for loop to work with
  the new type.

* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
- Add RetainPtrHashingCocoa[ARC].mm to the project.
* Tools/TestWebKitAPI/Tests/WTF/cf/RetainPtrHashing.cpp:
(TestWebKitAPI::TEST(RetainPtrHashing, HashSet)):
(TestWebKitAPI::TEST(RetainPtrHashing, HashMapKey)):
(TestWebKitAPI::TEST(RetainPtrHashing, HashMapValue)):
- Add more expectations to tests and fix header sort order.
* Tools/TestWebKitAPI/Tests/WTF/cocoa/RetainPtrHashingCocoa.mm: Copied from Tools/TestWebKitAPI/Tests/WTF/cf/RetainPtrHashing.cpp.
(TestWebKitAPI::TEST(RetainPtrHashingCocoa, HashSet)):
(TestWebKitAPI::TEST(RetainPtrHashingCocoa, HashMapKey)):
(TestWebKitAPI::TEST(RetainPtrHashingCocoa, HashMapValue)):
- Add tests for HashMap<RetainPtr<>> and HashSet<RetainPtr<>>.
* Tools/TestWebKitAPI/Tests/WTF/cocoa/RetainPtrHashingCocoaARC.mm: Add.
- Add tests for HashMap<RetainPtr<>> and HashSet<RetainPtr<>> under ARC.

Canonical link: https://commits.webkit.org/283028@main
